### PR TITLE
Gen4 Operator Refactoring: ORDER BY & Debugging Enhancements

### PR DESCRIPTION
**Pull request by systay**
## Description
Moved ORDER BY planning to operator phase, enabling arbitrary expressions
Added toTree method on operators for improved debugging

## Example query

To better understand the planning process for a single query, let's examine how we would plan the following query:

```sql
SELECT u.foo, ue.bar 
from user u join user_extra ue on u.uid = ue.uid 
ORDER BY u.baz
```

### Step 1
In the first step of planning, we have an operator tree that looks like this:

```
Horizon
└── ApplyJoin (u.uid = ue.uid)
    ├── Route (Scatter on user)
    │   └── Table (user.user)
    └── Route (Scatter on user)
        └── Filter (:u_uid = ue.uid)
            └── Table (user.user_extra)
```

Everything under a route will be turned into SQL and sent to MySQL.

The Horizon here represents the SELECT expressions and aggregations, ORDER BY, GROUP BY, LIMIT. If we can push this operator under a route,  we don't need to do any additional planning, and can just copy the query fields from the input query to the output query.

### Step 2
In the next step, we decided that we can't push the Horizon, and instead need to expand it into its components. 

```
Ordering (u.baz asc)
└── Projection (u.foo, ue.bar)
    └── ApplyJoin (u.uid = ue.uid)
        ├── Route (Scatter on user)
        │   └── Table (user.user)
        └── Route (Scatter on user)
            └── Filter (:u_uid = ue.uid)
                └── Table (user.user_extra)
```

The Horizon is split into an Ordering and a Projection operator.

### Step 3
We continue to push things down - the Projection is split and pushed to both sides of the join, and the Ordering is sent to the left side of the join.

```
ApplyJoin (u.uid = ue.uid)
├── Ordering (u.baz asc)
│   └── Projection (u.foo)
│       └── Route (Scatter on user)
│           └── Table (user.user)
└── Projection (ue.bar)
    └── Route (Scatter on user)
        └── Filter (:u_uid = ue.uid)
            └── Table (user.user_extra)
```

### Step 4
Finally, we are able to push both Projection and Ordering into the Route on the LHS of the join.

```
ApplyJoin (u.uid = ue.uid)
├── Route (Scatter on user)
│   └── Ordering (u.baz asc)
│       └── Projection (u.foo)
│           └── Table (user.user)
└── Route (Scatter on user)
    └── Projection (ue.bar)
        └── Filter (:u_uid = ue.uid)
            └── Table (user.user_extra)
```

So the vtgate plan is ultimately just a join, with one query sent to the left, and for each row we get from that, we will issue a query on the RHS of the join. The two queries are:

```sql
-- LHS
select u.foo, u.uid, u.baz, weight_string(u.baz) 
from `user` as u 
order by u.baz asc

-- RHS
select ue.bar from user_extra as ue where ue.uid = :u_uid
```

### Bonus

You might wonder why vtgate is asking MySQL to return the values for `u.baz` and the `weight_string()` for it, given that there doesn't seem to be any use of this column in the vtgate plan we end up with, right?

Well, since this Route query is a scatter query, we'll be getting results from multiple shards, and we have to merge-sort these results to get the final unified result. The data in this column is used by the Route operator to perform this merge-sort.

You might also ask, why are we still using weight_string? Don't we have good support for comparing values using the evalengine?

We do have good collation support, but it's not perfect. For this test, I used a fake database without sufficient schema information, so vtgate doesn't know the type of `u.baz`. To ensure that we can compare these values, we return both the raw column value and the weight_string value. Once the data arrives at vtgate, it can decide which of the two columns to sort by.

## Related Issue(s)
Tracking issue: https://github.com/vitessio/vitess/issues/11626

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by systay** [comment]
The first commit can also be found in https://github.com/vitessio/vitess/pull/12952

File: go/vt/vtgate/planbuilder/testdata/from_cases.json [hunk 1]
```diff
@@ -3398,32 +3407,40 @@
...
-                  "Sharded": true
-                },
-                "FieldQuery": "select 1 from user_extra where 1 != 1",
-                "Query": "select 1 from user_extra",
-                "Table": "user_extra"
+                "OperatorType": "Join",
+                "Variant": "Join",
+                "JoinColumnIndexes": "L:0,L:1",
+                "TableName": "`user`_user_extra",
+                "Inputs": [
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/from_cases.json:3420]: why do we have a simple projection on top of join? Can join handle the projection here?

File: go/vt/vtgate/planbuilder/testdata/memory_sort_cases.json [hunk 2]
```diff
@@ -606,49 +606,58 @@
...
           {
-            "OperatorType": "Join",
-            "Variant": "Join",
-            "JoinColumnIndexes": "L:1,L:2,R:0,R:1",
-            "JoinVars": {
-              "user_id": 0
-            },
-            "TableName": "`user`_music",
+            "OperatorType": "Sort",
+            "Variant": "Memory",
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/memory_sort_cases.json:618]: this looks like a slower plan

File: go/vt/vtgate/planbuilder/testdata/memory_sort_cases.json [hunk 3]
```diff
@@ -720,49 +729,58 @@
...
-            "OperatorType": "Join",
-            "Variant": "Join",
-            "JoinColumnIndexes": "L:1,L:2,R:0,L:3,R:1,L:4",
-            "JoinVars": {
-              "user_id": 0
-            },
-            "TableName": "`user`_music",
+            "OperatorType": "Sort",
+            "Variant": "Memory",
+            "OrderBy": "(3|4) ASC, (2|5) DESC, (1|6) ASC",
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/memory_sort_cases.json:742]: truncation at the end could avoid projection at operation.

File: go/vt/vtgate/planbuilder/testdata/union_cases.json [hunk 4]
```diff
@@ -2287,7 +2287,8 @@
   {
     "comment": "union with invalid order by clause with table qualifier",
     "query": "select id from user union select 3 order by id",
-    "plan": "VT12001: unsupported: ORDER BY on top of UNION"
+    "v3-plan": "VT12001: unsupported: ORDER BY on top of UNION",
+    "gen4-plan": "VT13001: [BUG] ORDER BY in complex query *planbuilder.distinct"
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/union_cases.json:2291]: instead of bug it should be unsupported error message.

**Comment by systay** [comment]
So, @harshit-gangal, I should have written this earlier. In an earlier PR, we did work to remove most of the unnecessary `SimpleProjection`s (#12790)

In this PR, I noticed that there are issues with how derived tables are being handled, which was breaking a lot of plans. So, I turned off the expanding of derived tables. If we can't push them down, we simply fail the operator planning and fall back on the old `logicalPlan` planner, and that is why you are seeing all these `SimpleProjection` again. 

My plan is to follow this PR up with another one where I re-enable the op-planner for derived tables.

File: go/vt/vtgate/planbuilder/testdata/memory_sort_cases.json (hunk 2 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/memory_sort_cases.json:618]: https://github.com/vitessio/vitess/pull/12954#issuecomment-1520004048

File: go/vt/vtgate/planbuilder/testdata/memory_sort_cases.json (hunk 3 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/memory_sort_cases.json:742]: https://github.com/vitessio/vitess/pull/12954#issuecomment-1520004048

File: go/vt/vtgate/planbuilder/testdata/from_cases.json (hunk 1 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/from_cases.json:3420]: https://github.com/vitessio/vitess/pull/12954#issuecomment-1520004048

File: go/vt/vtgate/planbuilder/testdata/union_cases.json (hunk 4 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/union_cases.json:2291]: I didn't quite figure out why this was happening. My thinking was that once we take on DISTINCT on the operators, we would figure out what's wrong.

File: go/vt/vtgate/planbuilder/operators/logical.go [hunk 5]
```diff
@@ -268,16 +267,17 @@ func getOperatorFromJoinTableExpr(ctx *plancontext.PlanningContext, tableExpr *s
 }
 
 func getOperatorFromAliasedTableExpr(ctx *plancontext.PlanningContext, tableExpr *sqlparser.AliasedTableExpr) (ops.Operator, error) {
+	tableID := ctx.SemTable.TableSetFor(tableExpr)
 	switch tbl := tableExpr.Expr.(type) {
 	case sqlparser.TableName:
-		tableID := ctx.SemTable.TableSetFor(tableExpr)
+		tableID := tableID
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/logical.go:273]: `tableID := tableID` this can probably be removed.

Comment (systay) [reply go/vt/vtgate/planbuilder/operators/logical.go:273]: image

**Review by harshit-gangal — APPROVED** [review]

**Review by arthurschreiber — APPROVED** [review]

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 6]
```diff
@@ -210,11 +214,15 @@ func (a *ApplyJoin) getJoinColumnFor(ctx *plancontext.PlanningContext, e sqlpars
 	return
 }
 
+func jcToAliasedExpr(column JoinColumn) sqlparser.Expr {
+	return column.Original.Expr
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:218]: same here but that's a nit

File: go/vt/vtgate/planbuilder/operators/horizon.go [hunk 7]
```diff
@@ -88,8 +86,34 @@ func (h *Horizon) src() ops.Operator {
...
+		return h.QP, nil
+	}
+	qp, err := CreateQPFromSelect(ctx, h.Select.(*sqlparser.Select))
+	if err != nil {
+		return nil, err
+	}
+	h.QP = qp
+	return h.QP, nil
+}
+func (h *Horizon) setQP(qp *QueryProjection) {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon.go:107]: new line too

**Review by frouioui — APPROVED** [review]

**Pull request merged**